From 5152429b89e97b46bd98978590693c664ae01b2c Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Mon, 20 Nov 2023 14:47:19 +0100
Subject: [PATCH 840/985] evl/proxy: introduce proxied polling operation

Allow for collecting events from the proxied file which might
implement any of the poll and oob_poll handlers.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/asm-generic/evl/poll.h | 31 ++++++++++++++++--
 include/evl/poll.h             | 37 ++--------------------
 include/evl/poll_head.h        | 48 ++++++++++++++++++++++++++++
 include/uapi/evl/control-abi.h |  2 +-
 kernel/evl/proxy.c             | 57 +++++++++++++++++-----------------
 5 files changed, 108 insertions(+), 67 deletions(-)
 create mode 100644 include/evl/poll_head.h

diff --git a/include/asm-generic/evl/poll.h b/include/asm-generic/evl/poll.h
index f7e15b75d..972698553 100644
--- a/include/asm-generic/evl/poll.h
+++ b/include/asm-generic/evl/poll.h
@@ -4,7 +4,7 @@
 
 #ifdef CONFIG_EVL
 
-#include <linux/list.h>
+#include <evl/poll_head.h>
 
 /*
  * Poll operation descriptor for f_op->oob_poll.  Can be attached
@@ -12,8 +12,6 @@
  */
 #define EVL_POLL_NR_CONNECTORS  4
 
-struct evl_poll_head;
-
 struct oob_poll_wait {
 	struct evl_poll_connector {
 		struct evl_poll_head *head;
@@ -24,11 +22,38 @@ struct oob_poll_wait {
 	} connectors[EVL_POLL_NR_CONNECTORS];
 };
 
+struct oob_poll_queue {
+	struct evl_poll_head head;
+};
+
+static inline
+void init_oob_poll_queue(struct oob_poll_queue *pwq)
+{
+	evl_init_poll_head(&pwq->head);
+}
+
+#define poll_signal_oob(__pwq, __mask)	\
+	evl_signal_poll_events(&(__pwq)->head, __mask)
+
+#define poll_wait_oob(__pwq, __wait)	\
+	evl_poll_watch(&(__pwq)->head, __wait, NULL)
+
 #else
 
 struct oob_poll_wait {
 };
 
+struct oob_poll_queue {
+};
+
+static inline void init_oob_poll_queue(struct oob_poll_queue *pwq) { }
+
+#define poll_signal_oob(__pwq, __mask)	\
+	do { (void)(__pwq), (void)(__mask); } while (0)
+
+#define poll_wait_oob(__pwq, __wait)	\
+	do { (void)(__pwq), (void)(__wait); } while (0)
+
 #endif	/* !CONFIG_EVL */
 
 #endif /* !_ASM_GENERIC_EVL_POLL_H */
diff --git a/include/evl/poll.h b/include/evl/poll.h
index fba3da8d9..c16ed26bb 100644
--- a/include/evl/poll.h
+++ b/include/evl/poll.h
@@ -12,6 +12,7 @@
 #include <linux/rbtree.h>
 #include <linux/spinlock.h>
 #include <linux/poll.h>
+#include <evl/poll_head.h>
 #include <evl/wait.h>
 #include <evl/factory.h>
 #include <uapi/evl/types.h>
@@ -19,18 +20,8 @@
 
 struct file;
 
-#define EVL_POLLHEAD_INITIALIZER(__name) {				\
-		.watchpoints = LIST_HEAD_INIT((__name).watchpoints),	\
-		.lock = __HARD_SPIN_LOCK_INITIALIZER((__name).lock),	\
-	}
-
-struct evl_poll_head {
-	struct list_head watchpoints; /* struct poll_watchpoint */
-	hard_spinlock_t lock;
-};
-
 struct evl_poll_node {
-	struct list_head next;	/* in evl_fd->poll_nodes */
+	struct list_head next;	/* in watchpoint->poll_nodes */
 };
 
 /*
@@ -47,30 +38,6 @@ struct evl_poll_watchpoint {
 	struct evl_poll_node node;
 };
 
-static inline
-void evl_init_poll_head(struct evl_poll_head *head)
-{
-	INIT_LIST_HEAD(&head->watchpoints);
-	raw_spin_lock_init(&head->lock);
-}
-
-void evl_poll_watch(struct evl_poll_head *head,
-		struct oob_poll_wait *wait,
-		void (*unwatch)(struct evl_poll_head *head));
-
-void __evl_signal_poll_events(struct evl_poll_head *head,
-			      int events);
-
-static inline void
-evl_signal_poll_events(struct evl_poll_head *head,
-		       int events)
-{
-	/* Quick check. We'll redo under lock */
-	if (!list_empty(&head->watchpoints))
-		__evl_signal_poll_events(head, events);
-
-}
-
 void evl_drop_poll_table(struct evl_thread *thread);
 
 void evl_drop_watchpoints(struct list_head *drop_list);
diff --git a/include/evl/poll_head.h b/include/evl/poll_head.h
new file mode 100644
index 000000000..253571f30
--- /dev/null
+++ b/include/evl/poll_head.h
@@ -0,0 +1,48 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (C) 2018 Philippe Gerum  <rpm@xenomai.org>
+ */
+
+#ifndef _EVL_POLL_HEAD_H
+#define _EVL_POLL_HEAD_H
+
+#include <linux/list.h>
+#include <linux/spinlock.h>
+
+struct oob_poll_wait;
+
+#define EVL_POLLHEAD_INITIALIZER(__name) {				\
+		.watchpoints = LIST_HEAD_INIT((__name).watchpoints),	\
+		.lock = __HARD_SPIN_LOCK_INITIALIZER((__name).lock),	\
+	}
+
+struct evl_poll_head {
+	struct list_head watchpoints; /* struct evl_poll_watchpoint */
+	hard_spinlock_t lock;
+};
+
+static inline
+void evl_init_poll_head(struct evl_poll_head *head)
+{
+	INIT_LIST_HEAD(&head->watchpoints);
+	raw_spin_lock_init(&head->lock);
+}
+
+void evl_poll_watch(struct evl_poll_head *head,
+		struct oob_poll_wait *wait,
+		void (*unwatch)(struct evl_poll_head *head));
+
+void __evl_signal_poll_events(struct evl_poll_head *head,
+			      int events);
+
+static inline void
+evl_signal_poll_events(struct evl_poll_head *head,
+		       int events)
+{
+	/* Quick check. We'll redo under lock */
+	if (!list_empty(&head->watchpoints))
+		__evl_signal_poll_events(head, events);
+}
+
+#endif /* !_EVL_POLL_HEAD_H */
diff --git a/include/uapi/evl/control-abi.h b/include/uapi/evl/control-abi.h
index 801e6e7dd..f9b933803 100644
--- a/include/uapi/evl/control-abi.h
+++ b/include/uapi/evl/control-abi.h
@@ -17,7 +17,7 @@
  * latter to the former. CAUTION: a litteral value is required for the
  * current ABI definition (scripts reading this may be naive).
  */
-#define EVL_ABI_LEVEL  34
+#define EVL_ABI_LEVEL  35
 
 #define EVL_CONTROL_DEV  "/dev/evl/control"
 
diff --git a/kernel/evl/proxy.c b/kernel/evl/proxy.c
index 35f07a2ad..d8d70fd78 100644
--- a/kernel/evl/proxy.c
+++ b/kernel/evl/proxy.c
@@ -11,6 +11,7 @@
 #include <linux/file.h>
 #include <linux/wait.h>
 #include <linux/fs.h>
+#include <linux/poll.h>
 #include <linux/log2.h>
 #include <linux/irq_work.h>
 #include <linux/workqueue.h>
@@ -60,12 +61,12 @@ struct evl_proxy {
 	struct evl_poll_head poll_head;
 };
 
-static inline bool proxy_is_readable(struct evl_proxy *proxy)
+static inline bool proxy_may_read(struct evl_proxy *proxy)
 {
 	return !!(proxy->element.clone_flags & EVL_CLONE_INPUT);
 }
 
-static inline bool proxy_is_writable(struct evl_proxy *proxy)
+static inline bool proxy_may_write(struct evl_proxy *proxy)
 {
 	return !!(proxy->element.clone_flags & EVL_CLONE_OUTPUT);
 }
@@ -399,7 +400,7 @@ static ssize_t proxy_oob_write(struct file *filp,
 	struct proxy_ring *ring = &proxy->output.ring;
 	ssize_t ret;
 
-	if (!proxy_is_writable(proxy))
+	if (!proxy_may_write(proxy))
 		return -ENXIO;
 
 	do {
@@ -421,7 +422,7 @@ static ssize_t proxy_oob_read(struct file *filp,
 	bool request_done = false;
 	ssize_t ret;
 
-	if (!proxy_is_readable(proxy))
+	if (!proxy_may_read(proxy))
 		return -ENXIO;
 
 	if (count == 0)
@@ -439,7 +440,7 @@ static ssize_t proxy_oob_read(struct file *filp,
 		ret = evl_wait_flag(&ring->oob_wait);
 		if (ret)
 			break;
-		if (atomic_cmpxchg(&in->on_eof, true, false) == true) {
+		if (atomic_cmpxchg(&in->on_eof, true, false)) {
 			ret = 0;
 			break;
 		}
@@ -457,20 +458,17 @@ static __poll_t proxy_oob_poll(struct file *filp,
 	__poll_t ret = 0;
 	int peek;
 
-	if (!(proxy_is_readable(proxy) || proxy_is_writable(proxy)))
-		return POLLERR;
-
 	evl_poll_watch(&proxy->poll_head, wait, NULL);
 
-	if (proxy_is_writable(proxy) &&
+	if (proxy_may_write(proxy) &&
 		atomic_read(&oring->fillsz) < oring->bufsz)
-		ret = POLLOUT|POLLWRNORM;
+		ret |= POLLOUT|POLLWRNORM;
 
 	/*
 	 * If the input ring is empty, kick the worker to perform a
-	 * readahead as a last resort.
+	 * readahead.
 	 */
-	if (proxy_is_readable(proxy)) {
+	if (proxy_may_read(proxy)) {
 		if (atomic_read(&iring->fillsz) > 0)
 			ret |= POLLIN|POLLRDNORM;
 		else if (atomic_read(&proxy->input.reqsz) == 0) {
@@ -480,6 +478,13 @@ static __poll_t proxy_oob_poll(struct file *filp,
 		}
 	}
 
+	/*
+	 * If the proxied file implements an oob poll handler, collect
+	 * the events pending there as well.
+	 */
+	if (proxy->filp->f_op->oob_poll)
+		ret |= proxy->filp->f_op->oob_poll(filp, wait);
+
 	return ret;
 }
 
@@ -490,7 +495,7 @@ static ssize_t proxy_write(struct file *filp, const char __user *u_buf,
 	struct proxy_ring *ring = &proxy->output.ring;
 	ssize_t ret;
 
-	if (!proxy_is_writable(proxy))
+	if (!proxy_may_write(proxy))
 		return -ENXIO;
 
 	do {
@@ -512,7 +517,7 @@ static ssize_t proxy_read(struct file *filp,
 	bool request_done = false;
 	ssize_t ret;
 
-	if (!proxy_is_readable(proxy))
+	if (!proxy_may_read(proxy))
 		return -ENXIO;
 
 	if (count == 0)
@@ -527,7 +532,7 @@ static ssize_t proxy_read(struct file *filp,
 			request_done = true;
 		}
 		relay_input(proxy);
-		if (atomic_cmpxchg(&in->on_eof, true, false) == true) {
+		if (atomic_cmpxchg(&in->on_eof, true, false)) {
 			ret = 0;
 			break;
 		}
@@ -543,25 +548,21 @@ static __poll_t proxy_poll(struct file *filp, poll_table *wait)
 	struct proxy_ring *iring = &proxy->input.ring;
 	__poll_t ret = 0;
 
-	if (!(proxy_is_readable(proxy) || proxy_is_writable(proxy)))
-		return POLLERR;
-
-	if (proxy_is_writable(proxy)) {
+	if (proxy_may_write(proxy)) {
 		poll_wait(filp, &oring->inband_wait, wait);
 		if (atomic_read(&oring->fillsz) < oring->bufsz)
 			ret = POLLOUT|POLLWRNORM;
 	}
 
-	if (proxy_is_readable(proxy)) {
+	if (proxy_may_read(proxy)) {
 		poll_wait(filp, &iring->inband_wait, wait);
 		if (atomic_read(&iring->fillsz) > 0)
 			ret |= POLLIN|POLLRDNORM;
-		else if (proxy->filp->f_op->poll) {
-			ret = proxy->filp->f_op->poll(proxy->filp, wait);
-			ret &= POLLIN|POLLRDNORM;
-		}
 	}
 
+	if (proxy->filp->f_op->poll)
+		ret |= proxy->filp->f_op->poll(proxy->filp, wait);
+
 	return ret;
 }
 
@@ -597,10 +598,10 @@ static int proxy_release(struct inode *inode, struct file *filp)
 {
 	struct evl_proxy *proxy = element_of(filp, struct evl_proxy);
 
-	if (proxy_is_writable(proxy))
+	if (proxy_may_write(proxy))
 		evl_flush_flag(&proxy->output.ring.oob_wait, EVL_T_RMID);
 
-	if (proxy_is_readable(proxy))
+	if (proxy_may_read(proxy))
 		evl_flush_flag(&proxy->input.ring.oob_wait, EVL_T_RMID);
 
 	return evl_release_element(inode, filp);
@@ -761,10 +762,10 @@ static void proxy_factory_dispose(struct evl_element *e)
 
 	proxy = container_of(e, struct evl_proxy, element);
 
-	if (proxy_is_writable(proxy))
+	if (proxy_may_write(proxy))
 		destroy_ring(&proxy->output.ring);
 
-	if (proxy_is_readable(proxy))
+	if (proxy_may_read(proxy))
 		destroy_ring(&proxy->input.ring);
 
 	fput(proxy->filp);
-- 
2.30.2

